x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java

```diff
@@ -63,6 +63,16 @@ public interface ExpressionEvaluator extends Releasable {
         /** A Factory for creating ExpressionEvaluators. */
         interface Factory {
             ExpressionEvaluator get(DriverContext context);
+
+            /**
+             * {@code true} if it is safe and fast to evaluate this expression non-lazily
+             * in contexts like {@code CASE}. This defaults to {@code false}, but expressions
+             * that evaluate quickly and can not produce warnings may override this to
+             * {@code true} to get a significant speed-up in {@code CASE}-like operations.
+             */
+            default boolean safeToEvalInLazy() {
+                return false;
+            }
```

Comment:
I *suppose* this is something that we could derive from the Expression - but it felt simpler to put it here. And it's just a boolean at this point - though maybe it should be a cost estimate at some point.

Reply:
In case of constants, the unvisited branches can be removed and the case simplified.

Reply:
That works if the expression is constant, but not if the *values* are constant. We still have to evaluate in that case. And when we do we can do it the fast way with something like this.